ddtrace/internal/wrapping.py

```diff
@@ -370,25 +375,40 @@ def wrap(f, wrapper):
     code.argnames = f.__code__.co_varnames[:nargs]
 
     f.__code__ = code.to_code()
-    wf = cast(WrapperFunction, f)
+    wf = cast(WrappedFunction, f)
     wf.__dd_wrapped__ = wrapped
 
     return wf
 
 
-def unwrap(wrapper):
-    # type: (WrapperFunction) -> FunctionType
+def unwrap(wf, wrapper):
+    # type: (WrappedFunction, Wrapper) -> FunctionType
     """Unwrap a wrapped function.
 
-    This is the reverse of :func:`wrap`.
+    This is the reverse of :func:`wrap`. In case of multiple wrapping layers,
+    this will unwrap the one that uses ``wrapper``. If the function was not
+    wrapped with ``wrapper``, it will return the first argument.
     """
     try:
-        wrapped = cast(FunctionType, wrapper.__dd_wrapped__)
-        assert wrapped, "Wrapper has wrapped function"
-        if wrapped.__name__ == "<wrapped>":
-            f = cast(FunctionType, wrapper)
-            f.__code__ = wrapped.__code__
-            del wrapper.__dd_wrapped__
-    except (IndexError, AttributeError):
-        return cast(FunctionType, wrapper)
-    return f
+        wrapped = cast(FunctionType, wf.__dd_wrapped__)
+
+        # Sanity check
+        assert wrapped.__name__ == "<wrapped>", "Wrapper has wrapped function"
+
+        if wrapper not in cast(FunctionType, wf).__code__.co_consts:
+            # This is not the correct wrapping layer. Keep looking
+            return unwrap(cast(WrappedFunction, wrapped), wrapper)
+
+        f = cast(FunctionType, wf)
+        f.__code__ = wrapped.__code__
+        try:
+            wf.__dd_wrapped__ = cast(WrappedFunction, wrapped).__dd_wrapped__  # type: ignore[assignment]
+        except AttributeError:
+            # No more wrapping layers. Restore the original function by removing
+            # this extra attribute.
+            del wf.__dd_wrapped__
+
+        return f
+
+    except AttributeError:
+        return cast(FunctionType, wf)
```

Comment:
I feel like this function needs more comments or a better description of the strategy

We check to see if `wf` has a `__dd_wrapped__` attribute, if not we just return `wf`.

Otherwise, we check to see if our wrapper exists in the constants for `wf`, if not, then we try to unwrap `wf.__dd_wrapped__` recursively until we either find the correct "layer" to unwrap, or there are no more layers.

If we we do have the correct wrapped layer, then we copy over the original function `__code__` to `wf`, and try to copy over any future wrappers?

Reply:
Correct. This is akin to removing an entry from a singly-linked list. Once we identify the element to remove, we need to update the link to by-pass the element we are removing, if any elements remain in the tail. I'll add some comments in the code.